### ISSUE TITLE
Retry failed requests

### ISSUE THREAD

**Issue by sindresorhus** [onevcat/Kingfisher#1424]
### Check List

- [x] I have read the [wiki page](https://github.com/onevcat/Kingfisher/wiki) and [cheat sheet](https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet), but there is no information I need.
- [x] I have searched in [existing issues](https://github.com/onevcat/Kingfisher/issues?utf8=✓&q=is%3Aissue), but did not find a same one.
- [x] I want to report a problem instead of asking a question. It'd better to use [kingfisher tag in Stack Overflow](http://stackoverflow.com/questions/tagged/kingfisher) to ask a question.

#### What

Networks are inherently unreliable. I use Kingfisher in SwiftUI to download a bunch of large images and I often experience failures, especially when using hotspot with 4G. It would be nice if Kingfisher could retry such errors. I'm actually surprised this is not already built into Kingfisher, as retrying is such a core feature when dealing with networks.

I imagine it would have some sane defaults, like 3 retries with exponential back-off, but be configurable if needed. I think it should retry on `.invalidHTTPStatusCode`, and also `.URLSessionError` if the error is `(error as NSError).code == NSURLErrorTimedOut`.

#### Reproduce

N/A

#### Other Comment

In https://github.com/onevcat/Kingfisher/issues/1058 you commented that you would be willing to look at this after Kingfisher 5.

You also argued:

> For me, since I always use Kingfisher in a table view or collection view, every time the cell be loaded, it will try to load the image. More or less, it behaves like a "retry" for downloading images.

But this is not the case when used in SwiftUI.

**Comment by onevcat** [comment]
Yes. A better retry mechanism is necessary, especially for the case in SwiftUI.

Before we can develop it, there is a bold "workaround" if you'd like to try, to use the `alternativeSources` option to download again when a failure happens:

```swift
extension KingfisherWrapper where Base: UIImageView {
    func setImageWithRetryThreeTimes(url: URL) {
        let source = Source.network(url)
        setImage(with: source, options: [.alternativeSources([source, source, source])])
    }
}

```

**Comment by sindresorhus** [comment]
I'm already using a similar workaround, although not as clever.

**Comment by onevcat** [comment]
The retry mechanism is already implemented in #1447 and it will be released soon.

To retry an image setting, pass a `.retryStrategy` option with an associated `RetryStrategy` value. Kingfisher provides a built-in `DelayRetryStrategy` with several parameters, like:

```swift
let retry = DelayRetryStrategy(maxRetryCount: 5, retryInterval: .seconds(3))
imageView.kf.setImage(with: imageURL, options: retry)
```

It will retry 5 times if there is something wrong in network, with a 3 seconds interval between each retry operation. You can also choose an `.accumulated(3)` way to have a `3 -> 6 -> 9 -> 12 -> 15` retry interval, or even define your own interval pattern by `.custom`.

If you need more control for the retry strategy, implement your own type conforming the `RetryStrategy` protocol and provide it as an option.

**Comment by onevcat** [comment]
Released with version 5.14.0.

**Comment by prostiak** [comment]
> Released with version 5.14.0.

Thank you. Didn't know it is here now. Saved a lot of time.

**Comment by smolskyaleksey** [comment]
Hello, Can I make that alternativeSources also apply retryStrategy.  Now I get only one try to download alternativeURL. 
I can add 3 times the same alternativeURL in alternativeSources, but is there maybe another way?
